doc/_scripts/software_maturity_scanner.py

```diff
@@ -0,0 +1,554 @@
+            all_boards.update(integration_platforms)
+
+    # Only use nRF boards, but not nRF51 series
+    all_boards = set(
+        filter(lambda b: b.startswith("nrf") and not b.startswith("nrf51"), all_boards)
+    )
+    all_socs = {b.split("_")[1] for b in all_boards}
+    return list(sorted(all_socs))
+
+
+def find_enclosing_parenthesis(rule: str) -> int:
+    """Get string length of a rule enclosed in parenthesis."""
+
+    level = 0
+    for i, char in enumerate(rule):
+        if char == "(":
+            level += 1
+        elif char == ")":
+            level -= 1
+        if level == 0:
+            return i + 1
+
+    logger.error(f"Invalid Kconfig rule '{rule}': parenthesis missing")
+
+
+def split_on(delimiter: str, rule: str) -> List[str]:
+    """Like regular split, but ignore delimiters in subrules."""
+
+    i = 0
+    parts = []
+    while i < len(rule):
+        if rule[i] == "(":
+            i += find_enclosing_parenthesis(rule[i:])
+        elif rule[i:].startswith(delimiter):
+            parts.append(rule[0:i].strip())
+            rule = rule[i + len(delimiter) :]
+            i = 0
+        else:
+            i += 1
+    parts.append(rule.strip())
+    return parts
+
+
+def parse_rule(rule: str) -> list:
+    """Parse a Kconfig rule and return a nested list of expressions.
+
+    A Kconfig rule can be described by the following CFG:
+    rule -> or-rule
+    or-rule -> and-rule | and-rule '||' or-rule
+    and-rule -> sub-rule | sub-rule '&&' and-rule
+    sub-rule -> symbol | '(' or-rule ')'
+    symbol -> any matches of the regular expression '[A-Z0-9_]+'
+
+    And-expressions have a higher precedence than or-expressions, making the
+    rule 'A || B && C' equivalent with 'A || (B && C)'
+
+    Each list in the returned nested list of expressions is prefixed by the
+    'AND' or 'OR' keyword.
+
+    For example, the rule 'A && (B || C)' will return the following list:
+    [OR, [AND, A, [OR, [AND, B], [AND, C]]]]
+    """
+
+    and_expressions = split_on("||", rule)
+    parsed_rule = ["OR"]
+    for expression in and_expressions:
+        symbols = split_on("&&", expression)
+        for i, symbol in enumerate(symbols):
+            if symbol.startswith("(") and symbol.endswith(")"):
+                symbols[i] = parse_rule(symbol[1:-1])
+            elif not re.match(r"^[A-Z0-9_]+$", symbol):
+                logger.error(f"Invalid Kconfig symbol '{symbol}'")
+            elif not symbol.startswith("CONFIG_"):
+                symbols[i] = "CONFIG_" + symbol
+        symbols.insert(0, "AND")
+        parsed_rule.append(symbols)
+    return parsed_rule
+
+
+def evaluate_rule(vars: Dict[str, kconfiglib.Variable], rule: list) -> bool:
+    """Evaluate a parsed Kconfig rule against a dictionary of Kconfig variables.
+
+    Returns:
+        False if any of the required variables are missing, True otherwise
+    """
+
+    assert rule[0] == "OR"
```

Comment:
for every assert, please log which rule is not valid

Reply:
The relevant rule is added with an f-string to the asserts